docs/proposals/train_api_proposal.md

```diff
@@ -0,0 +1,154 @@
+
+**<h3>Background</h3>**
+
+Currently, there are two flows for data scientists to start using Training operator for their distributed training needs. 
+
+**<h4>Traditional method</h4>**
+
+1. The end-user has to write a custom script that includes data fetching, model loading, and a training loop using pytorch libraries.
+2. Create an image and push it to a registry.
+3. Using the Kubeflow train sdk to create worker spec and master spec which uses the above image and using these two specs create a pytorch job.
+
+This method requires infrastructure knowledge including image creation, registry access and training job spec. This makes it difficult for data scientists to purely focus at the application layer.
+
+**<h4>Higher level SDK</h4>**
+
+To provide a better user experience, a new higher level SDK was added in[ https://github.com/kubeflow/training-operator/pull/1659](https://github.com/kubeflow/training-operator/pull/1659) which avoided the need of image creation. Instead of creating the image, we can directly pass the custom script to train_func parameter and final pytorch job is created
+
+```python
+training_client.create_job(
+   name=pytorchjob_name,
+   train_func=train_function,
+   num_worker_replicas=3, # How many PyTorch Workers will be created.
+)
+```
+
+The above method reduces the time taken but still the end user has to write the train_function method.
+
+In many cases like in LLM, training core code template is a standard and it doesn’t require any custom modification. For faster experimentation of such models, an easier api should exist to train/finetune models without writing custom Python code. HuggingFace Trainer class is a good example but it is not easy to set it up in a Kubernetes environment. Training operator can handle this complexity and provide a better abstracted layer to the user.
+
+**<h3>New Train API proposal:</h3>**
+
+```python
+From kubeflow.training import trainingClient
+# Arguments related to the model provider including credentials
+model_args = modelProviderClass()
+# Arguments related to the dataset provider including credentials
+dataset_args = datasetProviderClass()
+# Arguments related to the trainer code
+parameters = {key:value pairs}
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='provider://repo/model_name', 
+   dataset= 'provider://train_dataset_path',
+   eval_dataset = "provider://eval_dataset_path"
+   model_args, 
+   dataset_args, 
+   parameters
+)
+```
+
+Example: 
+
+```python
+@dataclass
+class HuggingFace:
+    access_token = field()
+    peft_config = field()
+    transformerClass = field()
+
+@dataclass
+class S3:
+    access_token = field()
+    region = field()
+ 
+trainingClient.train(
+   nodes=1, 
+   nprocs_per_node=4, 
+   model='hf://openchat/openchat_3.5', 
+   dataset= 's3://doc-example-bucket1/train_dataset',
```

Comment:
Do we need separate parameters for `model`, `dataset`, and `eval_dataset` if these values can be always part of ModelProvider and DatasetProvider ?

Reply:
Just to make the inputs similar to Hugging face Trainer class, I added it like this

Reply:
I see, but we already not really compatible with Trainer API since we are going to support various providers, isn't ?
Also, what if train_dataset and eval_dataset require different access tokens ?

Reply:
I agree model, dataset, and eval_dataset should be moved to ModelProvider and DatasetProvider respectively so that parameters are less and api is concise